### PR TITLE
DM-16053: Update write_header function

### PR DESCRIPTION
**Pull request by menanteau**

### CODE REVIEW

File: python/HeaderService/hslib.py [hunk 1]
```diff
@@ -21,6 +21,7 @@
 # Create a logger for all functions
 LOGGER = hutils.create_logger(level=logging.NOTSET, name='HEADERSERVICE')
 
+
```
Comment (timj) [inline python/HeaderService/hslib.py:24]: Whitespace/flake8 fixes should be in separate commits from functionality changes.

File: python/HeaderService/hutils.py [hunk 2]
```diff
@@ -384,17 +275,23 @@ def write_headers(self, filenames, MP=False, NP=2, md5=False):
...
+    def write_header_string(self, filename):
+        """ Write header as string"""
+        self.string_header()
+        with open(filename, 'w') as fobj:
+            fobj.write(self.hstring)
+
     def write_fits(self, filename, dtype='random', naxis1=None, naxis2=None, btype='int32'):
 
-        """ Write a dummy fits file -- use for testing only"""
+        """ Write a dummy fits file filles with random or zeros -- use for testing only"""
```
Comment (timj) [inline python/HeaderService/hutils.py]: "filled" rather than "filles"

File: python/HeaderService/hutils.py [hunk 3]
```diff
@@ -384,17 +275,23 @@ def write_headers(self, filenames, MP=False, NP=2, md5=False):
...
                 hdr = copy.deepcopy(self.header[extname])
                 fits.write(data, header=hdr, extname=extname)
 
+    def write_header_string(self, filename):
+        """ Write header as string"""
+        self.string_header()
+        with open(filename, 'w') as fobj:
+            fobj.write(self.hstring)
+
     def write_fits(self, filename, dtype='random', naxis1=None, naxis2=None, btype='int32'):
```
Comment (timj) [inline python/HeaderService/hutils.py]: Maybe clearer to name it as `write_dummy_fits` ?

File: python/HeaderService/hutils.py [hunk 4]
```diff
@@ -384,17 +275,23 @@ def write_headers(self, filenames, MP=False, NP=2, md5=False):
...
+        """Write a header file using the FITS format with empty HDUs"""
         data = None
         with fitsio.FITS(filename, 'rw', clobber=True, ignore_empty=True) as fits:
             for extname in self.HDRLIST:
                 hdr = copy.deepcopy(self.header[extname])
                 fits.write(data, header=hdr, extname=extname)
 
+    def write_header_string(self, filename):
+        """ Write header as string"""
+        self.string_header()
```
Comment (timj) [inline python/HeaderService/hutils.py]: It would look less like magic if the return value of this method was then printed out. `string_header()` would feel less magical I think if it was called `calculate_string_header()` so that it would more explicitly be updating `self`.  In this case it is much clearer to use the return value and not rely on the magic.

File: python/HeaderService/hutils.py [hunk 5]
```diff
@@ -384,17 +275,23 @@ def write_headers(self, filenames, MP=False, NP=2, md5=False):
...
+        with open(filename, 'w') as fobj:
+            fobj.write(self.hstring)
```
Comment (timj) [inline python/HeaderService/hutils.py]: I have a slight preference for `print(self.hstring, file=fobj, end='')` as being more expressive, but `write()` is allowed.

File: python/HeaderService/hutils.py [hunk 6]
```diff
@@ -427,19 +322,20 @@ def write_fits(self, filename, dtype='random', naxis1=None, naxis2=None, btype='
...
+        t0 = time.time()
+        if self.write_mode == 'fits':
+            self.write_header_fits(filename)
+        elif self.write_mode == 'string':
+            self.write_header_string(filename)
         else:
-            t0 = time.time()
-            self.write_header_emptyHDU(filename)
-            LOGGER.info("Header write time:{}".format(elapsed_time(t0)))
+            LOGGER.info("ERROR: header write_mode:{} not recognized".format(self.write_mode))
```
Comment (timj) [inline python/HeaderService/hutils.py]: This seems like more of a `LOGGER.error` or `LOGGER.warning` rather than `LOGGER.info`.  Given this should never happen and given that you want a header written should it not raise?

**Review by timj — APPROVED** [review]
I have some minor comments. Also please start to use numpydoc docstrings as you add/modify methods.

File: python/HeaderService/hslib.py (hunk 1 above)
Comment (menanteau) [reply python/HeaderService/hslib.py:24]: Will separate these commits.

File: python/HeaderService/hutils.py (hunk 2 above)
Comment (menanteau) [reply python/HeaderService/hutils.py]: fixed

File: python/HeaderService/hutils.py (hunk 3 above)
Comment (menanteau) [reply python/HeaderService/hutils.py]: Changes to `write_dummy_fits` as suggested.

File: python/HeaderService/hutils.py (hunk 6 above)
Comment (menanteau) [reply python/HeaderService/hutils.py]: Yes, it should be `LOGGER.error(message)` followed by a `raise ValueError(msg)`

File: python/HeaderService/hutils.py (hunk 4 above)
Comment (menanteau) [reply python/HeaderService/hutils.py]: Changed to function name to `calculate_string_header()` and use the return value rather than the `self` for clarity.

File: python/HeaderService/hutils.py (hunk 5 above)
Comment (menanteau) [reply python/HeaderService/hutils.py]: will keep the `write()` method as I prefer it.
